src/utilities/downloads_backend.jl

```diff
@@ -0,0 +1,109 @@
+function get_downloader(downloader=nothing)
+    lock(AWS_DOWNLOAD_LOCK) do
+        yield() # let other downloads finish
+        downloader isa Downloader && return
+        while true
+            downloader = AWS_DOWNLOADER[]
+            downloader isa Downloader && return
+            D = Downloader()
+            D.easy_hook = (easy, info) -> Curl.setopt(easy, Curl.CURLOPT_FOLLOWLOCATION, false)
+            AWS_DOWNLOADER[] = D
+        end
+    end
+    return downloader
+end
+
+# https://github.com/JuliaWeb/HTTP.jl/blob/2a03ca76376162ffc3423ba7f15bd6d966edff9b/src/MessageRequest.jl#L84-L85
+body_length(x::AbstractVector{UInt8}) = length(x)
+body_length(x::AbstractString) = sizeof(x)
+
+read_body(x::IOBuffer) = take!(x)
+read_body(x::IO) = readavailable(x)
+
+function AWS._http_request(backend::DownloadsBackend, request)
+    # If we pass `output`, Downloads.jl will expect a message
+    # body in the response. Specifically, it sets
+    # <https://curl.se/libcurl/c/CURLOPT_NOBODY.html>
+    # only when we do not pass the `output` argument.
+    #
+    # When the method is `HEAD`, the response may have a Content-Length
+    # but not send any content back (which appears to be correct,
+    # <https://stackoverflow.com/a/18925736/12486544>).
+    #
+    # Thus, if we did not set `CURLOPT_NOBODY`, and it gets a Content-Length
+    # back, it will hang waiting for that body.
+    #
+    # Therefore, we do not pass an `output` when the `request_method` is `HEAD`.
+    if request.response_stream === nothing
+        request.response_stream = IOBuffer()
+    end
+    output_arg = if request.request_method == "HEAD"
+        NamedTuple()
+    else
+        (; output=request.response_stream)
+    end
+
+    # If we're going to return the stream, we don't want to read the body into an
+    # HTTP.Response we're never going to use. If we do that, the returned stream
+    # will have no data available (and reading from it could hang forever).
+    body_arg = if request.request_method == "HEAD" || request.return_stream
+        () -> NamedTuple()
+    else
+        () -> (; body = read_body(request.response_stream))
+    end
+
+    # HTTP.jl sets this header automatically.
+    request.headers["Content-Length"] = string(body_length(request.content))
+
+    # We pass an `input` only when we have content we wish to send.
+    input = IOBuffer()
+    input_arg = if !isempty(request.content)
+        write(input, request.content)
+        (; input=input)
+    else
+        NamedTuple()
+    end
+
+    @repeat 4 try
+        downloader = @something(backend.downloader, get_downloader())
+        # set the hook so that we don't follow redirects. Only
+        # need to do this on per-request downloaders, because we
+        # set our global one with this hook already.
+        if backend.downloader !== nothing
+            downloader.easy_hook = (easy, info) -> Curl.setopt(easy, Curl.CURLOPT_FOLLOWLOCATION, false)
+        end
```

Comment:
This could be a thread-safety issue if multiple threads are using the same downloader. Though I think multiple threads using the same downloader is already an issue upstream.

This maybe makes things slightly better but doesn't actually fix the race:
```suggestion
        if backend.downloader !== nothing && downloader.easy_hook === nothing
            downloader.easy_hook = (easy, info) -> Curl.setopt(easy, Curl.CURLOPT_FOLLOWLOCATION, false)
        end
```
Not totally sure what we should do here.

Reply:
@ericphanson I think we need per-request hooks to fix this. See https://github.com/JuliaLang/Downloads.jl/issues/141